### ISSUE TITLE
BluetoothGattCallback.OnCharacteristicWrite not called on OS prior to 8.0

### ISSUE THREAD

**Issue by onesadjam** [xamarin/xamarin-android#1843]
### Steps to Reproduce

1. Use BluetoothDevice.ConnectGatt to connect to a known bluetooth device
2. OnConnectionStateChange to connected, make call to BluetoothGatt.DiscoverServices
3. OnServicesDiscovered success, identify expected service and characteristic.  Call BluetoothGattCharacteristic.SetValue to assign a new value, then BluetoothGatt.WriteCharacteristic(characteristicToWrite) to send the request to the device.

### Expected Behavior
Once BluetoothGatt.WriteCharacteristic is called the OnCharacteristicWrite callback should be called to either indicate the value was successfully written to the device or the write failed.

### Actual Behavior
When the target platform is running Android OS 8.0 or higher the behavior is as expected.  If running an earlier version of the OS the callback is never called.

### Version Information

Microsoft Visual Studio Professional 2017 
Version 15.7.3
VisualStudio.15.Release/15.7.3+27703.2026
Microsoft .NET Framework
Version 4.7.03056

Installed Version: Professional

Microsoft Visual Studio Tools for Applications 2017   00369-60000-00001-AA736
Microsoft Visual Studio Tools for Applications 2017

Application Insights Tools for Visual Studio Package   8.12.10405.1
Application Insights Tools for Visual Studio

ASP.NET and Web Tools 2017   15.0.40522.0
ASP.NET and Web Tools 2017

ASP.NET Core Razor Language Services   15.7.31476
Provides languages services for ASP.NET Core Razor.

ASP.NET Web Frameworks and Tools 2012   4.0.21208.0
For additional information, visit https://www.asp.net/

ASP.NET Web Frameworks and Tools 2017   5.2.60419.0
For additional information, visit https://www.asp.net/

Azure App Service Tools v3.0.0   15.0.40424.0
Azure App Service Tools v3.0.0

Azure Functions and Web Jobs Tools   15.0.40424.0
Azure Functions and Web Jobs Tools

C# Tools   2.8.3-beta6-62923-07. Commit Hash: 7aafab561e449da50712e16c9e81742b8e7a2969
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Fabric.DiagnosticEvents   1.0
Fabric Diagnostic Events

JavaScript Language Service   2.0
JavaScript Language Service

JetBrains ReSharper Ultimate 2017.3.3    Build 111.0.20180302.65130
JetBrains ReSharper Ultimate package for Microsoft Visual Studio. For more information about ReSharper Ultimate, visit http://www.jetbrains.com/resharper. Copyright © 2018 JetBrains, Inc.

Merq   1.1.19-rc (a4ffc1b)
Command Bus, Event Stream and Async Manager for Visual Studio extensions.

Microsoft Azure Service Fabric Tools for Visual Studio   2.1
Microsoft Azure Service Fabric Tools for Visual Studio

Microsoft Azure Tools   2.9
Microsoft Azure Tools for Microsoft Visual Studio 2017 - v2.9.10420.2

Microsoft Continuous Delivery Tools for Visual Studio   0.3
Simplifying the configuration of continuous build integration and continuous build delivery from within the Visual Studio IDE.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual Studio Tools for Containers   1.1
Develop, run, validate your ASP.NET Core applications in the target environment. F5 your application directly into a container with debugging, or CTRL + F5 to edit & refresh your app without having to rebuild the container.

Mono Debugging for Visual Studio   4.10.5-pre (ab58725)
Support for debugging Mono processes with Visual Studio.

NuGet Package Manager   4.6.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit http://docs.nuget.org/.

ProjectServicesPackage Extension   1.0
ProjectServicesPackage Visual Studio Extension Detailed Info

ResourcePackage Extension   1.0
ResourcePackage Visual Studio Extension Detailed Info

SlowCheetah - XML config transforms   1.0
This package enables you to transform your app.config or any other XML file based on the build configuration. It also adds additional tooling to help you create XML transforms.

SQL Server Data Tools   15.1.61804.210
Microsoft SQL Server Data Tools

SQL Server Integration Services   
Microsoft SQL Server Integration Services Designer
Version 14.0.800.98

TypeScript Tools   15.7.20419.2003
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   2.8.3-beta6-62923-07. Commit Hash: 7aafab561e449da50712e16c9e81742b8e7a2969
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual F# Tools 10.1 for F# 4.1   15.7.0.0.  Commit Hash: 2527e6829ecdc8281ee60d83be8cfd0fa720a648.
Microsoft Visual F# Tools 10.1 for F# 4.1

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

VisualStudio.Mac   1.0
Mac Extension for Visual Studio

Xamarin   4.10.10.1 (f1760154c)
Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

Xamarin Designer   4.12.1 (f3257e429)
Visual Studio extension to enable Xamarin Designer tools in Visual Studio.

Xamarin.Android SDK   8.3.3.2 (HEAD/dffc59120)
Xamarin.Android Reference Assemblies and MSBuild support.

Xamarin.iOS and Xamarin.Mac SDK   11.12.0.4 (64fece5)
Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.

**Comment by JonDouglas** [comment]
Hi @onesadjam,

Thanks for reporting an issue. Can you please attach a [minimal, complete, and verifiable example](https://github.com/xamarin/xamarin-android/wiki/Submitting-Bugs,-Feature-Requests,-and-Pull-Requests#bug-reports) to demonstrate the problem?

From the description of this issue, this seems more like an Android API behavior rather than a problem with Xamarin.Android. Another way to confirm this would be to create a Java / Kotlin Android application with the same code and see if the behavior is the same. If the behavior is the same, it would suggest a limitation with these APIs in your implementation.

There are a few similar Stack Overflow questions that may help as well:

https://stackoverflow.com/questions/30420137/writecharacteristic-returns-true-but-does-not-call-oncharacteristicwrite

https://stackoverflow.com/questions/17910322/android-ble-api-gatt-notification-not-received

https://stackoverflow.com/questions/20043388/working-with-ble-android-4-3-how-to-write-characteristics

**Comment by onesadjam** [comment]
@JonDouglas 
Thanks for the response.  I have two apps I'm working with.  One is provided by the vendor of the BLE device I'm working with.  We'll call that one VendorApp.  It is native Java code for Android.  The other app I have is the one I'm developing myself, XamApp. I have multiple phones to work with, ranging in OS version from 5 to 8.1.0.  What I observe is that if I use the VendorApp (which I have verified with the vendor is making the same calls I make in XamApp) the attempt to write a characteristic using the sequence setValue(); writeCharacteristic() results in a callback to onCharacteristicWrite().  Using the XamApp, I receive the same behavior, but only when deployed to devices running OS 8.0 or higher.  When I deploy to 7.0 or earlier, the onCharacteristicWrite() callback is never called.  I've been trying a variety of things, and I have verified that the onCharacteristicRead() callback does work correctly, regardless of the OS on the device.

Given my observation that XamApp works as expected on 8.0 and later, but not earlier OS versions and that VendorApp works on any OS 5.0 or higher, my conclusion was that the error is not in the system API but in the Xamarin layer above it.

Thank you for the SO links as well.  Unfortunately, I was already familiar with them as part of my debugging effort.  The first link describes my issues very well.  Unfortunately, I've verified that I'm supplying the correct number of bytes when setting the value of the characteristic, and also given that the same APK succeeds on OS 8.0 and higher gives me some confidence that my code is correct and the issue must lie elsewhere.

**Comment by JonDouglas** [comment]
@onesadjam Thanks for the background on the issue. I believe we will need to investigate this a bit further.

Can we get a [minimal, complete, and verifiable example](https://github.com/xamarin/xamarin-android/wiki/Submitting-Bugs,-Feature-Requests,-and-Pull-Requests#bug-reports) for both of these scenarios? (Xamarin.Android vs. Java/Kotlin). In other words, can you upload two projects that use the exact same code which demonstrates the `OnCharacteristicWrite()` callback not being called on 7.0 and lower for Xamarin.Android and is being called on 7.0 and lower for Java/Kotlin?

Please also include specifics regarding the devices you're testing on since Android can be a fickle beast and we will want to best reproduce your scenario locally to see what's going on.

**Comment by onesadjam** [comment]
@JonDouglas I'll do my best to get code to you with reproduction of the issue, but for the time being I can at least give you the battery of devices I've been testing against:


Make | Model | OS | Java | Xamarin
-- | -- | -- | -- | --
Google | Pixel 2 | 8.1.0 | PASS | PASS
Google | Pixel XL | 8.1.0 | PASS | PASS
Samsung | S9 Plus | 8.0.0 | PASS | PASS
Samsung | Note 5 | 7.0 | PASS | FAIL
Samsung | Note 3 | 5.0 | PASS | FAIL
Samsung | S7 | 8.0.0 | PASS | PASS
Motorola | DROID Turbo | 6.0.1 | PASS | FAIL
Samsung | Galaxy S8 | 8.0.0 | PASS | PASS

**Comment by JonDouglas** [comment]
@onesadjam Thank you kindly! This is very helpful.

**Comment by onesadjam** [comment]
@JonDouglas I've created a minimal, complete, and verifiable example which you can find here: https://github.com/onesadjam/BleGattTest

This is the Xamarin side, which exhibits the issue when I deploy to 8.0+ succeeds in writing the characteristic and when deployed to a lower revision fails to receive the OnCharacteristicWrite callback.

I'll try to get you a working example built in Android Studio that exhibits working in both cases.

I'm curious as well to know if there may be some particular nuance of my development environment (versions of SDKs loaded?) that could be an underlying culprit as well.  I've attached a snapshot of the project properties page as viewed from my environment in case that may be related.

![androidmanifestsettings](https://user-images.githubusercontent.com/1125927/41624866-f4439400-73e4-11e8-87d7-3dcf44cc7462.JPG)
![applicationsettings](https://user-images.githubusercontent.com/1125927/41624867-f451e596-73e4-11e8-9f0e-f1f76411a859.JPG)

**Comment by onesadjam** [comment]
@JonDouglas I think I've gotten to the bottom of this and I don't believe it has anything to do with Xamarin Android.  I ported the Xamarin version of code above to Java and found that I was getting the same behavior: success on 8.0 and 8.1, failure on anything else.  I spent the day trying a variety of things, and in the end I found the root issue was the time to establish a good bond with the bluetooth device.  Here are some of my notes on the changes necessary to make the process reliable:

- Need to pair to the device 
- After pairing to the device, the first connection attempt may be disconnected, so be prepared to reconnect
- After connecting, attempting to write to a characteristic may not get an onCharacteristicWrite response. Instead, use readCharacteristic or readDescriptor which will trigger their callbacks
- Attempts to readCharacteristic or readDescriptor may indicate failure in the callback method due to GATT_AUTH_FAIL. In this case, retry the read attempt until it is successful.
- Once the read is successful attempts to write will also trigger the onCharacteristicWrite callback.  It is safe to make writes now.

I suspect that my 8.0 and 8.1 devices are fast enough that they are negotiating the pairing / authentication much faster than my older devices, and I was simply lucky not to run into these gotchas.
